addons/purchase_stock/models/purchase.py

```diff
@@ -103,6 +103,14 @@ def button_cancel(self):
                         if all(state in ('done', 'cancel') for state in siblings_states):
                             move_dest_ids.write({'procure_method': 'make_to_stock'})
                             move_dest_ids._recompute_state()
+                        if order_line.propagate:
+                            move_dest_ids._action_cancel()
+            if order.state in ('purchase'):
```

Comment:
You don't need to do 2 loops for order depending the state. You could always cancel the line's move_ids if they exists and manage the move_dest_ids depending the propagate field.